### PR TITLE
[tt-train] SDPA FP32 logsumexp intermediates

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Ticket
Link to Github Issue

### Problem description
The SDPA forward kernel previously stored two BF16 intermediate tensors (max_value and recip_sum_exp, each 64 tiles wide) that were passed to the backward kernels for softmax recomputation. This BF16 precision was a source of numerical error that compounded through the backward pass, contributing to training loss spikes. Additionally, the ring attention merge logic (Context Parallelism) required separate global_max and global_sum_exp tracking, leading to complex and numerically fragile online softmax accumulation.

#### FP32 logsumexp intermediate format:

- SDPA forward now emits a single FP32 logsumexp (LSE) tensor (32 tiles wide, down from 2x 64 BF16 tiles) defined as lse = max + log(sum_exp)
- Both backward kernels (sdpa_bw_q, sdpa_bw_kv) consume the FP32 LSE directly and apply fused softmax recomputation on DST registers via apply_softmax_statistics_on_dst
- Dataflow kernels updated to read/write single FP32 intermediate tile per row instead of two BF16 tiles

#### SDPA forward precision improvements:

- cb_attention_weights (QK result) buffer promoted to FP32
- cb_prev_mm_out / cb_cur_mm_out (V accumulation) buffers promoted to FP32 with UnpackToDestFp32 mode
- Output update (update_cur_mm_out) rewritten to use SFPU operations for full FP32 precision in DST registers
- compute_and_pack_lse helper added for FP32 LSE computation
- Block size capped at 3 when fp32_dest_acc_en=true to prevent DST register overflow (4 FP32 registers, need block_size + 1 for unary_bcast)
- Deferred scaling: scale factor applied after max-subtraction in apply_exp_inplace_and_find_exp_sum instead of during masking

#### SDPA backward precision improvements:

- Intermediate CB switched from BF16 to FP32
- Q kernel: UnpackToDestFp32 enabled for attention weights CB (full FP32 precision, no transpose needed)
- KV kernel: attention weights CB uses FP32 data format with Default unpack mode (TF32 through SrcA), uses FPU transpose. SFPU transpose_wh_dest is broken on Blackhole, so FPU path is required for cross-architecture compatibility.

#### Ring attention (ring_attention_sdpa.cpp):

- Forward merge rewritten from max/recip_sum_exp tracking to logaddexp-based accumulation — simpler and numerically stable
- Backward pass weight computation simplified to exp(lse_j - final_lse)
- global_lse initialized to -infinity (FP32)

#### Ring SDPA device operations:

- ring_sdpa_fw_device_operation.cpp: intermediate shape updated from 64 to 32, dtype from query.dtype() to FLOAT32

#### Tests:

- Float reference updated to produce single lse intermediate
- Tolerance adjusted to 3e-2 for D=128 configs (inherent BF16 precision gap over longer matmul inner dimension)
- New NIGHTLY_RingAttentionMergeSimulation test: single-device simulation of 4-device ring attention verifying logaddexp merge logic for both forward and backward passes

#### N300 (single card) TinyLlama (1B) training:

- Baseline (BF16 intermediates) | 1932.06 ms
- FP32 intermediates only | 1911.93 ms
- FP32 intermediates + FP32 sdpa_fw buffers + TF32 KV attention weights | 1932.67 ms

The FP32 intermediate format itself is perf-neutral (or slightly faster due to reduced tile count: 1 FP32 tile vs 2 BF16 tiles per row). The ~20 ms regression comes from promoting cb_attention_weights to FP32, which is the price for improved numerical stability. This will be recovered once SFPU row-reduce-max and sub_bcast_col primitives become available from the LLK team, enabling full FP32 softmax in DST without the extra pack/unpack overhead. Please take a look here: https://github.com/tenstorrent/tt-metal/issues/41686

### Checklist

- [x] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates)
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates)
- [x] New/Existing tests provide coverage for changes
- [x] (Optional) Ran [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/runs/24273148454) on the PR branch to catch linting errors early

### CI Status
_Auto-generated on every push. Badges update live. Click a pipeline name or badge to filter by this branch._

| Pipeline | Status | Latest Run |
|---|:---:|---|
| [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [#2545](https://github.com/tenstorrent/tt-metal/actions/runs/24268966540) |
| [Blackhole post-commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) |
| [L2 Nightly](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [#5110](https://github.com/tenstorrent/tt-metal/actions/runs/24268982382) |
| [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) |
| [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) |
| [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) | [Launch ↗](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:vmelnykov/sdpa_fp32_logsumexp_intermediates) |

### CODE REVIEW

File: tt-train/tests/ops/sdpa_bw_op_test.cpp [hunk 1]
```diff
@@ -816,3 +854,290 @@ TEST_F(SDPABackwardTest, NIGHTLY_CausalMask_LargerSequence) {
...
+//
+// Verifies:
+// 1. FP32 logsumexp intermediates are correct
+// 2. The logaddexp merge reproduces full-sequence SDPA output
+// 3. The backward pass gradient scaling via exp(lse_j - final_lse) is correct
+
+TEST_F(SDPABackwardTest, NIGHTLY_RingAttentionMergeSimulation) {
+    GTEST_SKIP()
+        << "Single-device ring attention simulation — requires manual opt-in, not part of regular nightly runs";
+    using namespace ttml;
```
[the remark this answers is not in the document]
Comment (vmelnykovTT) [reply tt-train/tests/ops/sdpa_bw_op_test.cpp:821]: This test was added as a quick manual sanity check for SDPA kernel behavior inside composite Ring Attention, especially when a multi-device VM isn’t available. It’s not intended for CI — it should only be used for convenient manual testing while iterating on further SDPA optimizations/changes.

**Review by dyurshevichTT — APPROVED** [review]
ring sdpa works with the change

**Review by zbaczewskiTT — APPROVED** [review]
Great job! Looks good.
Thank you, I've learned something.
